### PR TITLE
[WIP] Add ModeLeave, ModeEnter events

### PR DESCRIPTION
**Pull request by nimitbhardwaj**
Add the ModeLeave and ModeEnter events to be used with autocmd

### CODE REVIEW

**Comment by nimitbhardwaj** [comment]
This PR is now in WIP, I would like to know if my plan about this PR is correct, I am planning that, on entering a mode, 2 events will be triggered, ModeEnter and ModeLeave, because, to get to some mode we have to leave some other mode, this info will be stored in the vim variables (variables which are used with `:v`), there is dictionary `vimvardict` in `globals.h`, I will put 2 entities in it first is `from`, other is `to`, names are easy to be understood, I check there are 7 modes, namely `Normal`, `Visual`, `Select`, `Operator-Pending`, `Insert`, `Command-Line` and `Terminal`, events for each mode will be created.
Autocmd will be like that 
`:autocmd ModeEnter <mode listed here> <cmd>`, mode listed means if event occured is ModeEnter, and we enter in the listed mode, then cmd will take place

**Comment by nimitbhardwaj** [comment]
This PR, only now takes care of visual, insert and normal mode, may be I would need help to get the places where to trigger the events to carry this PR.

File: src/nvim/ex_getln.c [hunk 1]
```diff
@@ -394,6 +394,7 @@ static uint8_t *command_line_enter(int firstc, long count, int indent)
     try_enter(&tstate);
     apply_autocmds(EVENT_CMDLINELEAVE, (char_u *)firstcbuf, (char_u *)firstcbuf,
                    false, curbuf);
+    apply_autocmds(EVENT_MODELEAVE, (char_u *)"cmdline", NULL, false, curbuf);
```
Comment (nimitbhardwaj) [inline src/nvim/ex_getln.c]: This event didn't trigger, on `autocmd ModeLeave cmdline echom 'Hello'` or on even `autocmd ModeLeave * echom`, but the command `autocmd CmdlineLeave * echom` executes, I even tried by changing this line to `apply_autocmds(EVENT_MODELEAVE, (char_u *)firstcbuf, (char_u *)firstcbuf, false, curbuf);`, but it didn't work

Comment (nimitbhardwaj) [reply src/nvim/ex_getln.c]: But rest of Events triggered like `autocmd ModeLeave insert echom 'Hello'`  works fine

**Comment by nimitbhardwaj** [comment]
Added the basic modes, but there may be some mistakes, especially in normal mode, comments are welcomed

File: src/nvim/edit.c [hunk 2]
```diff
@@ -275,7 +275,10 @@ static void insert_enter(InsertState *s)
 
     set_vim_var_string(VV_INSERTMODE, (char *) s->ptr, 1);
     set_vim_var_string(VV_CHAR, NULL, -1);
+
     apply_autocmds(EVENT_INSERTENTER, NULL, NULL, false, curbuf);
+    apply_autocmds(EVENT_MODEENTER, (char_u *)"insert", NULL, false, curbuf);
+    apply_autocmds(EVENT_MODELEAVE, (char_u *)"norm", NULL, false, curbuf);
```
Comment (bfredl) [inline src/nvim/edit.c]: I don't think we want pairs like this, modes in nvim is more like a stack machine than a finite state machine. Like, when pressing `:` in normal mode, then cmdline mode is entered. Similarly, when `<c-f>` is pressed in cmdline mode, again normal mode is entered. But when `<esc>` is pressed in cmdline mode, the mode is instead left (i e popped of the stack). If we want to publish both old and new mode, use `v:event` is better.

File: src/nvim/ex_getln.c [hunk 3]
```diff
@@ -369,6 +369,8 @@ static uint8_t *command_line_enter(int firstc, long count, int indent)
 
     apply_autocmds(EVENT_CMDLINEENTER, (char_u *)firstcbuf, (char_u *)firstcbuf,
                    false, curbuf);
+    apply_autocmds(EVENT_MODEENTER, (char_u *)"cmdline", NULL, false, curbuf);
+    apply_autocmds(EVENT_MODELEAVE, (char_u *)"norm", NULL, false, curbuf);
```
Comment (bfredl) [inline src/nvim/ex_getln.c]: also why not `"normal"`

Comment (nimitbhardwaj) [reply src/nvim/ex_getln.c]: yes normal will also be fine

File: src/nvim/edit.c (hunk 2 above)
Comment (nimitbhardwaj) [reply src/nvim/edit.c]: Hmm, I didn't had much idea of it, OK sure, I understand it, I will edit the code and will soon make a PR again.
Till now, I again make this PR as WIP

**Comment by justinmk** [comment]
will covered by this vim patch: https://github.com/vim/vim/commit/f1e8876fa2359b572d262772747405d3616db670

**Pull request closed without merging**